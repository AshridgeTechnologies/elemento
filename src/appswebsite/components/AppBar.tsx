import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}