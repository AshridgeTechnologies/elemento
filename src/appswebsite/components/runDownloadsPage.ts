import thePage from './DownloadsPage'
import {run} from '../../appsShared/renderInPage'

run(thePage)
