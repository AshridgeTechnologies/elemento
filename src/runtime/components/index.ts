export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
