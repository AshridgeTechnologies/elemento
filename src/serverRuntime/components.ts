export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts