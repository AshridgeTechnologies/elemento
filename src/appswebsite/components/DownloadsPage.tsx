import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Online - Downloads' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <Container>
                    <Stack maxWidth={600} spacing={3} mt={3}>
                        <Typography variant='h4'>Elemento Development Server</Typography>
                        <Typography>Please download the correct version for your operating system.</Typography>
                        <Typography>Your computer will complain that the app is not known, approved, signed etc.<br/>
                            We do not currently have the resources to jump through all the hoops required to get the app
                            approved on each operating system,
                            so you will just have to trust us here, and continue past all the warnings.
                        </Typography>
                        <Link href='/devServer/win/ElementoDevServer.exe'><Typography variant='h5'>Windows</Typography></Link>
                        <Link href='/devServer/macos/ElementoDevServer.zip'><Typography variant='h5'>Mac</Typography></Link>
                        <Link href='/devServer/linux/ElementoDevServer'><Typography variant='h5'>Linux</Typography></Link>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
