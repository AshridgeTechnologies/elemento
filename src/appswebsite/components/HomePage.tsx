import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import HomeIcon from '@mui/icons-material/Home'
import LightbulbIcon from '@mui/icons-material/LightbulbTwoTone'
import CreateIcon from '@mui/icons-material/CreateTwoTone'
import AodIcon from '@mui/icons-material/AodTwoTone'
import ArrowCircleRightTwoToneIcon from '@mui/icons-material/ArrowCircleRightTwoTone';

const primary = theme.palette.primary.main
const secondary = theme.palette.secondary.main

function Logo() {
    return <Box color={secondary} border='8px solid' borderRadius={4} width={108} height={104} paddingTop='10px' paddingX='3px'>
        <Stack>
            <Typography fontSize='100px' fontFamily='"Helvetica"' align='center' lineHeight={0.8}>El</Typography>
            <Typography fontSize='16px' lineHeight={1}>42</Typography>
        </Stack>
    </Box>
}

function SectionLink(props: {title: string, href: string, icon: any, children: any}) {
    const Icon = props.icon
    return <Box border='4px solid' borderColor={primary} borderRadius={4} width='300px' paddingY='10px' paddingX='16px'>
        <Link href={props.href} underline='none'>
            <Stack spacing={2}>
                <Box textAlign='center'><Icon sx={{ fontSize: 180 }} /></Box>
                <Link component='div' underline='hover' href={props.href} variant='h5' textAlign='center'>{props.title}</Link>
                <Typography fontSize='20px' color='#555'>{props.children}</Typography>
                <Box textAlign='center'><ArrowCircleRightTwoToneIcon sx={{ fontSize: 32, color: secondary }} /></Box>
            </Stack>
        </Link>
    </Box>
}

export default function HomePage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Home'/>
            </Box>
            <Box flex='1' minHeight={0} mt={2}>
                <Container maxWidth="lg">
                    <Stack spacing={2}>
                        <Stack direction='row'>
                            <Logo/>
                            <Stack ml={4}>
                                <Typography fontWeight='bold' variant='h3'  color={secondary} align='center'>Elemento</Typography>
                                <Typography variant='h4' color={secondary} align='center' mt={2}>
                                    Everything you need to make and use your own software
                                </Typography>
                            </Stack>
                        </Stack>
                        <Stack direction='row' justifyContent='space-between' alignItems='stretch' marginTop='30px'>
                            <SectionLink title='Learn' href='/learn' icon={LightbulbIcon}>
                                A great range of resources to teach you how to get started and create your first app with Elemento.
                            </SectionLink>
                            <SectionLink title='Create' href='/studio' icon={CreateIcon}>
                                Put your app together and preview it live in Elemento Studio
                            </SectionLink>
                            <SectionLink title='Use' href='/run' icon={AodIcon}>
                                Run your app - or go back to the apps you have already made.
                            </SectionLink>
                        </Stack>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
