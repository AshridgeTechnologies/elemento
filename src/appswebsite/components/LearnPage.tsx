import React from 'react'
import Toolbar from '@mui/material/Toolbar'
import {Box, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import {TutorialRunner} from '../../tutorials/TutorialRunner'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Learn' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <TutorialRunner/>
            </Box>
        </Box>
    </ThemeProvider>
}
