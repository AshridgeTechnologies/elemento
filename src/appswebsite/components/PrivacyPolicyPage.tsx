import React from 'react'
import {Box, Stack, Typography} from '@mui/material'
import AppBar from './AppBar'

export default function PrivacyPolicyPage() {

    return <Box display='flex' flexDirection='column' height='100%' width='100%'>
        <Box flex='0'>
            <AppBar/>
        </Box>
        <Box flex='1' minHeight={0}>
            <Stack spacing={2}>
                <Typography variant='h4'>Elemento Apps - Privacy Policy</Typography>
                <Typography variant='body1'>Coming soon</Typography>
            </Stack>
        </Box>
    </Box>
}
