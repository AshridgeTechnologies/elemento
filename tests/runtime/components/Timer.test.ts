/**
 * @vitest-environment jsdom
 */
import {expect, test, vi} from 'vitest'
import {BaseComponentState, Timer} from '../../../src/runtime/components/index'
import {createStateFn, valueObj, wait, wrappedTestElement} from '../../testutil/testHelpers'
import {render} from '@testing-library/react'

import {testContainer} from '../../testutil/rtlHelpers'

const [timer, appStoreHook] = wrappedTestElement(Timer)
const stateAt = (path: string) => appStoreHook.stateAt(path)

const createState = createStateFn(Timer.State)
const {store} = createState
const getLatest = <T extends BaseComponentState<any, any>>(state: T): T => store.get(state._path!)

// Start of possible technique for mocking time, but also uses Date AND requestAnimationFrame only runs the callback once
// let rafCallback: FrameRequestCallback
// globalThis.requestAnimationFrame = (callback: FrameRequestCallback) => {
//     rafCallback = callback
//     return 1
// }
//
// const runAnimationFrames = (times: number) => {
//     for (let i = 0; i < times; i++) rafCallback(0)
// }

test('Timer element produces output with properties supplied', () => {
    const {container} = render(timer('app.page1.timer', {
        period: 10,
        styles: {width: 23},
        label: 'Time Left'
    }))
    expect(container.innerHTML).toMatchSnapshot()
})

test('Timer element produces output with default values where properties omitted',
    () => {
        const {container} = render(timer('app.page1.timer', {}))
        expect(container.innerHTML).toMatchSnapshot()
    }
)

test.skip('Timer shows elapsed time value', async () => {
    const {el} = testContainer(timer('app.page1.timer1', {show: true, period: 0.2, interval: 0.1}))
    stateAt('app.page1.timer1').Start()
    await wait(120)
    expect(el`app.page1.timer1`.textContent).toBe('0.10')
})

test('State class has correct input properties', () => {
    const intervalFn = ()=> 10, endFn = ()=> 20
    const state = new Timer.State({period: 10, interval: 2, intervalAction: intervalFn, endAction: endFn })
    expect(state.period).toBe(10)
    expect(state.interval).toBe(2)
    expect(state.intervalAction).toBe(intervalFn)
    expect(state.endAction).toBe(endFn)
})

test('State class has correct input properties from value objects', () => {
    const state = new Timer.State({period: valueObj(10), interval: valueObj(2) })
    expect(state.period).toBe(10)
    expect(state.interval).toBe(2)
})

test('State class has null input properties if not supplied', () => {
    const state = new Timer.State({})
    expect(state.period).toBe(null)
    expect(state.interval).toBe(null)
    expect(state.intervalAction).toBe(null)
    expect(state.endAction).toBe(null)
})

test('Has correct output properties before started', () => {
    const state = new Timer.State({period: 10, interval: 2 })
    expect(state.startTime).toBe(null)
    expect(state.elapsedTime).toBe(0)
    expect(state.intervalTime).toBe(0)
    expect(state.remainingTime).toBe(10)
    expect(state.intervalCount).toBe(0)
    expect(state.isStarted).toBe(false)
    expect(state.isRunning).toBe(false)
    expect(state.isFinished).toBe(false)
})

test('Has correct output properties immediately after started', () => {
    const state = createState({period: 2, interval: 0.05 })
    const startTime = new Date()
    state.Start()
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime(), -1)
    expect(state.elapsedTime).toBeCloseTo(0, 2)
    expect(state.intervalTime).toBe(0)
    expect(state.remainingTime).toBeCloseTo(2, 2)
    expect(state.intervalCount).toBe(0)
    expect(state.isStarted).toBe(true)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
})

test('Has correct output properties immediately after started without fixed period', () => {
    const state = createState({ interval: 0.05, intervalAction: vi.fn(), endAction: vi.fn() })
    const startTime = new Date()
    state.Start()
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    expect(state.intervalTime).toBe(0)
    expect(state.elapsedTime).toBeCloseTo(0, 2)
    expect(state.remainingTime).toBe(null)
    expect(state.intervalCount).toBe(0)
    expect(state.isStarted).toBe(true)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
})

test('Calls interval action while running', async () => {
    const intervalAction = vi.fn()
    const state = createState({period: 2, interval: 0.03, intervalAction })
    const startTime = new Date()
    state.Start()
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    await wait(80)  // 2 intervals
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    expect(state.intervalCount).toBe(2)
    expect(state.intervalTime).toBe(0.06)
    expect(state.elapsedTime).toBeCloseTo(0.08, 2)
    expect(state.value).toBeCloseTo(0.08, 2)
    expect(state.remainingTime).toBeCloseTo(1.92, 2)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(2)
    expect(intervalAction).toHaveBeenLastCalledWith(state)
})

test('Has correct output properties if start called while running', async () => {
    const intervalAction = vi.fn()
    const state = createState({period: 2, interval: 0.03, intervalAction })
    const startTime = new Date()
    state.Start()
    await wait(80)  // 2 intervals

    state.Start()

    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    expect(state.intervalCount).toBe(2)
    expect(state.intervalTime).toBe(0.06)
    expect(state.elapsedTime).toBeCloseTo(0.08, 1)
    expect(state.remainingTime).toBeCloseTo(1.92, 1)
    expect(state.isStarted).toBe(true)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(2)
    expect(intervalAction).toHaveBeenLastCalledWith(state)
})

test('Has correct output properties while running without fixed period ', async () => {
    const state = createState({interval: 0.03, intervalAction: vi.fn(), endAction: vi.fn() })
    const startTime = new Date()
    state.Start()
    await wait(80)  // 2 intervals
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    expect(state.intervalCount).toBe(2)
    expect(state.intervalTime).toBe(0.06)
    expect(state.elapsedTime).toBeCloseTo(0.082, 2)
    expect(state.remainingTime).toBe(null)
    expect(state.isStarted).toBe(true)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
})

test('Has correct output properties while running without fixed interval or period', async () => {
    const intervalAction = vi.fn()
    const state = createState({intervalAction, endAction: vi.fn() })
    const startTime = new Date()
    state.Start()
    await wait(36)
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    expect(state.intervalCount).toBeGreaterThanOrEqual(2)
    expect(state.intervalCount).toBeLessThanOrEqual(3)
    expect(state.intervalTime).toBe(null)
    expect(state.elapsedTime).toBeCloseTo(0.038, 2)
    expect(state.remainingTime).toBe(null)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(state.intervalCount)
})

test('Has correct output properties while running with period but no fixed interval', async () => {
    const intervalAction = vi.fn()
    const state = createState({period: 3, intervalAction, endAction: vi.fn() })
    const startTime = new Date()
    state.Start()
    await wait(40)
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime(), 0)
    const intervalCount = state.intervalCount
    expect(intervalCount).toBeGreaterThanOrEqual(2) // if requestAnimationFrame is 60 fps
    expect(intervalCount).toBeLessThanOrEqual(3)
    expect(state.intervalTime).toBe(null)
    expect(state.remainingTime).toBeCloseTo(2.966, 1)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(intervalCount)
})

test('Has correct output properties and calls to actions when finished', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({period: 0.09, interval: 0.03, intervalAction, endAction })
    const startTime = new Date()
    state.Start()
    await wait(120)  // should be finished
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime(), 0)
    expect(state.intervalCount).toBe(2) // last interval comes just after finished
    expect(state.intervalTime).toBe(0.06)
    expect(state.elapsedTime).toBeCloseTo(0.09, 2)
    expect(state.remainingTime).toBe(0)
    expect(state.isStarted).toBe(false)
    expect(state.isRunning).toBe(false)
    expect(state.isFinished).toBe(true)
    const stateStartTime = state.startTime?.getTime() as number
    expect(state.finishedTime?.getTime()).toBeGreaterThanOrEqual(stateStartTime + 90)
    expect(state.finishedTime?.getTime()).toBeLessThanOrEqual(stateStartTime + 120)
    expect(intervalAction).toHaveBeenCalledTimes(2)
    expect(endAction).toHaveBeenCalledTimes(1)
    expect(endAction).toHaveBeenLastCalledWith(state)
})

test('Has correct output properties and calls to actions when finished with incomplete interval', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({period: 0.075, interval: 0.03, intervalAction, endAction })
    const startTime = new Date()
    state.Start()
    await wait(95)  // 2 intervals and finished
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    const intervalCount = state.intervalCount // test is flaky unless allow for missed raf calls
    expect(intervalCount).toBeGreaterThanOrEqual(1)
    expect(intervalCount).toBeLessThanOrEqual(2)
    expect(state.intervalTime).toBe(0.03 * intervalCount)
    expect(state.isStarted).toBe(false)
    expect(state.isFinished).toBe(true)
    expect(state.isRunning).toBe(false)
    expect(state.elapsedTime).toBe(0.075)
    expect(state.remainingTime).toBe(0)
    expect(intervalAction).toHaveBeenCalledTimes(intervalCount)
    // expect(intervalAction).toHaveBeenLastCalledWith(updatedState)
    expect(endAction).toHaveBeenCalledTimes(1)
    expect(endAction).toHaveBeenLastCalledWith(state)
})

test('Has correct output properties and calls to actions when finished with no fixed interval', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({period: 0.06, intervalAction, endAction })
    const startTime = new Date()
    state.Start()
    await wait(85)  // finished
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    const intervalCount = state.intervalCount
    expect(intervalCount).toBeGreaterThanOrEqual(3)
    expect(intervalCount).toBeLessThanOrEqual(4)
    expect(state.intervalTime).toBe(null)
    expect(state.remainingTime).toBe(0)
    expect(state.isRunning).toBe(false)
    expect(state.isStarted).toBe(false)
    expect(state.isFinished).toBe(true)
    expect(intervalAction).toHaveBeenCalledTimes(intervalCount)
    expect(endAction).toHaveBeenCalledTimes(1)
    expect(endAction).toHaveBeenLastCalledWith(state)
})

test('Has correct output properties and calls after Stop', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({interval: 0.03, intervalAction, endAction })
    const startTime = new Date()
    state.Start()
    await wait(80)  // 2 intervals
    const stateBeforeStop = getLatest(state)
    stateBeforeStop.Stop()
    expect(state.startTime?.getTime()).toBeCloseTo(startTime.getTime())
    expect(state.intervalCount).toBe(2)
    expect(state.intervalTime).toBe(0.06)
    expect(state.elapsedTime).toBeCloseTo(0.08, 2)
    expect(state.remainingTime).toBe(null)
    expect(state.isStarted).toBe(true)
    expect(state.isRunning).toBe(false)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(2)
    expect(intervalAction).toHaveBeenLastCalledWith(stateBeforeStop)
    expect(endAction).not.toHaveBeenCalled()

    await wait(30)  //make sure nothing else happens
    expect(intervalAction).toHaveBeenCalledTimes(2)
    expect(endAction).not.toHaveBeenCalled()
    expect(state).toStrictEqual(state)
    expect(state.elapsedTime).toBeCloseTo(0.08, 2)
})

test('Stop, Stop again and Start without Reset', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({interval: 0.05, period: 1.0, intervalAction, endAction })
    const startTime = new Date()
    state.Start()
    await wait(115)  // 2 intervals

    state.Stop()
    await wait(30)

    state.Stop()
    await wait(30)

    state.Start()
    await wait(70)  // at least one more interval

    const stateAfterRestart = getLatest(state)
    const newStartTime = stateAfterRestart.startTime?.getTime()
    expect(newStartTime).toBeCloseTo(startTime.getTime() + 175, -1)
    expect(stateAfterRestart.intervalCount).toBe(3)
    expect(stateAfterRestart.intervalTime).toBeCloseTo(0.15, 2)
    expect(stateAfterRestart.elapsedTime).toBeCloseTo(0.185, 2)
    expect(stateAfterRestart.remainingTime).toBeCloseTo(1.0 - 0.185, 2)
    expect(stateAfterRestart.isStarted).toBe(true)
    expect(stateAfterRestart.isRunning).toBe(true)
    expect(stateAfterRestart.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(3)
    expect(intervalAction).toHaveBeenLastCalledWith(stateAfterRestart)
    expect(endAction).not.toHaveBeenCalled()
})

test('Stop, Start again without Reset', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({interval: 0.05, period: 1.0, intervalAction, endAction })
    const startTime = new Date()
    state.Start()
    await wait(115)  // 2 intervals

    state.Stop()
    await wait(30)

    expect(state.intervalCount).toBe(2)

    state.Start()
    await wait(70)  // at least one more interval

    const newStartTime = state.startTime?.getTime()
    expect(newStartTime).toBeCloseTo(startTime.getTime() + 145, -1)
    expect(state.intervalCount).toBe(3)
    expect(state.intervalTime).toBeCloseTo(0.15, 2)
    expect(state.elapsedTime).toBeCloseTo(0.185, 2)
    expect(state.remainingTime).toBeCloseTo(1.0 - 0.185, 2)
    expect(state.isStarted).toBe(true)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(3)
    expect(intervalAction).toHaveBeenLastCalledWith(state)
    expect(endAction).not.toHaveBeenCalled()
})

test('Stop, Reset, Start', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({interval: 0.03, intervalAction, endAction })
    state.Start()
    await wait(80)  // 2 intervals

    state.Stop()
    await wait(20)

    expect(intervalAction).toHaveBeenCalledTimes(2)
    state.Reset()

    const restartTime = new Date()
    state.Start()
    await wait(50) // 1 interval

    const newStartTime = state.startTime?.getTime()
    expect(newStartTime).toBeCloseTo(restartTime.getTime(), -1)
    expect(state.intervalCount).toBe(1)
    expect(state.intervalTime).toBeCloseTo(0.03, 2)
    expect(state.elapsedTime).toBeCloseTo(0.05, 2)
    expect(state.remainingTime).toBe(null)
    expect(state.isStarted).toBe(true)
    expect(state.isRunning).toBe(true)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(3)
    expect(intervalAction).toHaveBeenLastCalledWith(state)
    expect(endAction).not.toHaveBeenCalled()
})

test('Reset while running', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({interval: 0.03, intervalAction, endAction })
    state.Start()
    await wait(50)  // 1 interval

    state.Reset()
    await wait(20)

    expect(state.startTime).toBe(null)
    expect(state.intervalCount).toBe(0)
    expect(state.intervalTime).toBe(0)
    expect(state.elapsedTime).toBe(0)
    expect(state.remainingTime).toBe(null)
    expect(state.isStarted).toBe(false)
    expect(state.isRunning).toBe(false)
    expect(state.isFinished).toBe(false)
    expect(intervalAction).toHaveBeenCalledTimes(1)
    expect(endAction).not.toHaveBeenCalled()
})

test('Has correct output properties and calls to actions when finished after Stop and re-Start', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({period: 0.15, interval: 0.06, intervalAction, endAction })
    state.Start()
    await wait(70)
    state.Stop()
    await wait(100)
    const restartTime = new Date()
    state.Start()
    await wait(50)
    expect(state.isFinished).toBe(false)
    await wait(70)

    expect(state.startTime?.getTime()).toBeCloseTo(restartTime.getTime(), -1)
    expect(state.intervalCount).toBe(2)
    expect(state.intervalTime).toBeCloseTo(0.12, 2)
    expect(state.elapsedTime).toBeCloseTo(0.15, 2)
    expect(state.remainingTime).toBe(0)
    expect(state.isStarted).toBe(false)
    expect(state.isRunning).toBe(false)
    expect(state.isFinished).toBe(true)
    expect(intervalAction).toHaveBeenCalledTimes(2)
    expect(endAction).toHaveBeenCalledTimes(1)
})

test('Has correct output properties and calls to actions after Start, finish, Start again', async () => {
    const intervalAction = vi.fn(), endAction = vi.fn()
    const state = createState({period: 0.15, interval: 0.06, intervalAction, endAction })
    state.Start()
    await wait(200)
    state.Start()
    await wait(200)

    expect(state.intervalCount).toBe(2)
    expect(state.intervalTime).toBeCloseTo(0.12, 2)
    expect(state.elapsedTime).toBeCloseTo(0.15, 2)
    expect(state.remainingTime).toBe(0)
    expect(state.isRunning).toBe(false)
    expect(state.isStarted).toBe(false)
    expect(state.isFinished).toBe(true)
    expect(endAction).toHaveBeenCalledTimes(2)
    expect(intervalAction).toHaveBeenCalledTimes(4)
})
