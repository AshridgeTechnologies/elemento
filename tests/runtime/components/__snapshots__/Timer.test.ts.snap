// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer element produces output with default values where properties omitted 1`] = `"<div class="MuiFormControl-root MuiTextField-root css-1qr9tsw-MuiFormControl-root-MuiTextField-root"><div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"><input aria-invalid="false" id="app.page1.timer" type="text" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input" value="0.00"><fieldset aria-hidden="true" class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"><legend class="css-ihdtdm"><span class="notranslate">â€‹</span></legend></fieldset></div></div>"`;

exports[`Timer element produces output with properties supplied 1`] = `"<div class="MuiFormControl-root MuiTextField-root css-c212v0-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root" data-shrink="true" for="app.page1.timer" id="app.page1.timer-label">Time Left</label><div class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"><input aria-invalid="false" id="app.page1.timer" type="text" class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input" value="0.00"><fieldset aria-hidden="true" class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"><legend class="css-14lo706"><span>Time Left</span></legend></fieldset></div></div>"`;
